### PR TITLE
KAFKA-13409: Catch cases where we attempt to terminate the VM in a junit test

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/utils/Exit.java [hunk 1]
```diff
@@ -30,16 +33,35 @@
...
+
+    static void throwIfInJunitTest() {
+        for (StackTraceElement[] elements : Thread.getAllStackTraces().values()) {
+            for (StackTraceElement element : elements) {
+                if (JUNIT_CLASSS_NAMES.contains(element.getClassName())) {
+                    throw new RuntimeException("Attempted to terminate the VM in a junit test.");
+                }
+            }
+        }
+    }
```
Comment (soarez) [inline clients/src/main/java/org/apache/kafka/common/utils/Exit.java:51]: Is the intention to allow termination to happen from other threads even if still running a test?
i.e. do we want this test to pass as well?

```scala
  @Test
  def exitWithoutSettingUpHooks(): Unit = {
    val t = new Thread {
      override def run {
        try {
          Exit.exit(1, None)
        } catch {
          case e: RuntimeException => assertEquals("Attempted to terminate the VM in a junit test.", e.getMessage)
        }
      }
    }
    t.start()
    t.join()
  }
```

It's also maybe slightly nicer to not have to check the stack frames

```suggestion
    private static final boolean IN_JUNIT_TEST = Stream.of(
            "org.junit.platform.commons.util.ReflectionUtils",
            "org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor"
    ).anyMatch(clazz -> {
        try {
            Class.forName(clazz);
            return true;
        } catch (ClassNotFoundException ignored) {
            return false;
        }
    });

    static void throwIfInJunitTest() {
        if (IN_JUNIT_TEST) {
            throw new RuntimeException("Attempted to terminate the VM in a junit test.");
        }
    }
```

Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/utils/Exit.java:51]: > Is the intention to allow termination to happen from other threads even if still running a test?

Why would we want this?

> (proposed changes)

It seems like somebody could end up with the junit jars on their CLASSPATH, but not be running a test. So I think looking at the stack trace is safer.

**Comment by cmccabe** [comment]
Here is an example of halt() being called from a junit test:
```
java.lang.RuntimeException: TRIED TO CALL HALT
    at org.apache.kafka.common.utils.Exit$1.execute(Exit.java:42)
    at org.apache.kafka.common.utils.Exit.halt(Exit.java:100)
    at kafka.utils.Exit$.halt(Exit.scala:33)
    at kafka.log.LogManager.handleLogDirFailure(LogManager.scala:201)
    at kafka.server.ReplicaManager.handleLogDirFailure(ReplicaManager.scala:1922)
    at kafka.server.ReplicaManager$LogDirFailureHandler.doWork(ReplicaManager.scala:246)
    at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:96)
```

**Comment by cmccabe** [comment]
For now, I'm going to close this since #11457 solves the case of tests derived from `QuorumTestHarness`. Maybe we'll need to revisit later if alternate integration test harnesses gain more ground or pure unit tests start hitting this.

**Pull request closed without merging**